### ISSUE TITLE
Compute transitive closures for GAV/GA/G graphs

### ISSUE THREAD

**Issue by jensdietrich** [nkiru-ede/MavenNetworkStudy#12]
Add additional processing step (with tests, readme description etc) to compute the TC of those three graphs, create plots for vertex and edge counts.

**Comment by nkiru-ede** [comment]
done @jensdietrich

**Comment by jensdietrich** [comment]
@nkiru-ede I just reviewed this -- some questions: 
1.  the code in `/Transitive_deps.py` looks pretty complex, did you write this from scratch ? Is there a lib that can do this (I think [boost](https://www.boost.org/doc/libs/1_85_0/libs/graph/doc/transitive_closure.html) can be used from python - link has TC function) ?
2. I cannot see any tests -- we **always** need them ! 
3. the script seems to operate on the input graph - we would need this using any directed graph as input so that we can compute this for GA and G graphs as well

**Comment by nkiru-ede** [comment]
yes, it looks complex because I firstly identified and removed cycles/loops before using topological sorting to get the transitive dependencies (this is done to a depth on 10) - as I was getting hundreds of millions of relationships otherwise as it kept going into endless loops.

There are algorithms such as Floyd-Warshall, Tarjan, etc. I tried them but they were very memory intensive and kept crashing because of the size of the graph. 

jensdietrich

**Comment by jensdietrich** [comment]
we need tests for this -- if you first collapse SCCs this is ok, then break this up into two scripts (one computes SCCs, the other one TC) and then test those separately

**Comment by nkiru-ede** [comment]
@jensdietrich I have broken down the scripts:

1. Remove loops and Cycle - [Remove Loops and Cycle](https://github.com/nkiru-ede/MavenNetworkStudy/blob/main/Project/RemoveLoops_Cycles.py), [Test](https://github.com/nkiru-ede/MavenNetworkStudy/blob/main/Project/test_RemoveLoops_Cycles.py)
2. Identify Sccs - [Identifying SCCs](https://github.com/nkiru-ede/MavenNetworkStudy/blob/main/Project/SCC.py), [Test](https://github.com/nkiru-ede/MavenNetworkStudy/blob/main/Project/SCC_test.py)

**Comment by jensdietrich** [comment]
Please create tests for each step now

**Comment by nkiru-ede** [comment]
@jensdietrich I included the links to the tests as well. But I need to move them to right folder and test that they are working on the two os. Will drop a message here once it is done

**Comment by nkiru-ede** [comment]
@jensdietrich  done

**Comment by jensdietrich** [comment]
@nkiru-ede can you please update the instructions on the main README -- they look wrong (in particular step 3)

**Comment by nkiru-ede** [comment]
@jensdietrich 

**FOR GAV**

Total count of Transitive dependencies (without condensation) - 11, 664, 867 

Total count of Transitive dependencies (with condensation)  -  11, 664, 867 

Find the file here - https://drive.google.com/file/d/1--XUMovC1k_5AAm6eM0Mi_vJMOPkRUZW/view?usp=sharing 

 

**Depth**  

The highest node depth is 35 

 
<img width="128" alt="1" src="https://github.com/user-attachments/assets/d4b1458c-146e-47c5-9cd0-3436e426d1c0">
<img width="93" alt="image" src="https://github.com/user-attachments/assets/1f3aebbd-70e1-429b-b076-84c3a96ea376">
![image](https://github.com/user-attachments/assets/c261cc49-a3a5-4dd6-b433-d75ebd6f3b0b)

**Transitive dependencies(Vertex and Edges):**

<img width="512" alt="image" src="https://github.com/user-attachments/assets/3e1432c3-950a-4d6b-b196-61c7c26461c1">

**Comment by nkiru-ede** [comment]
@jensdietrich computing transitive dependencies for GA is proving difficult as there as so many cycles that need to be removed first. I am still working on this

**Comment by jensdietrich** [comment]
@nkiru-ede   remember that SCC detection is fast (linear -- Tarjan's algorithm).

**Comment by nkiru-ede** [comment]
@jensdietrich Overall, condensing SCCs in GAV did not affect the transitive dependencies. I earlier shared the vertex and edge counts of the transitive dependencies computation of GAV.

Below are those GA  and G   - I first condensed the SCCs in both before computing the transitive dependencies, as both contained many cycles.

<img width="458" alt="image" src="https://github.com/user-attachments/assets/d83070a0-5266-4e03-af11-94baed8058e2">
<img width="382" alt="image" src="https://github.com/user-attachments/assets/57d8d35b-845e-4c50-833c-c1b15b96777f">

**Comment by jensdietrich** [comment]
check the readme -- Step 3: Compute Transitive Closures for GAV/GA/G looks wrong .
